### PR TITLE
[:hammer:] add chapter AP, Bridge & BCC

### PR DESCRIPTION
**Pull request by Lotanie**
- [x] Read [Contribute to NTHU-CPP](https://nthu-cp.github.io/NTHU-CPP/others/contribution.html).
- [x] Rebase to the latest main branch.
- [x] List all the references.
- [x] Successfully build the website by mdBook with no error after the change.
- [x] Exclude any irrelevant files.
- [x] Link to the issue which is related to your change.
- [x] Review the content by myself.
- [x] Pass tests/CI.

### CODE REVIEW

**Comment by Lotanie** [comment]
@harry900831 想請助教幫忙看看大綱與文章架構(題解的部分還在進行中)

File: src/SUMMARY.md [hunk 1]
```diff
@@ -21,6 +21,8 @@
   - [杜教篩](math/du_sieve.md)
   - [Revisit Arithmetic Function](math/revisit_arithmetic_function.md)
 - [Miscellaneous]()
+- [Graph]()
```
Comment (harry900831) [inline src/SUMMARY.md:24]: move this section above to the place that under dynamic programming

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+
+我們可以看個例子。
+
+<img src="image/DFS Tree.gif" width="400" style="display:block; margin: 0 auto;"/>
+
+如果把上圖畫成 DFS tree 就會長成如下圖這樣。我們發現在 DFS tree 上，**一個點可以通過 back edge，回到他的祖先**。
+
+<img src="image/DFS Tree.JPG" width="200" style="display:block; margin: 0 auto;"/>
+
+在無向圖上做 DFS tree 時要注意的是: 無向圖的 DFS 會讓一條邊被看到 2 次。如果你的 code 寫的是看到一條邊就根據它是 tree edge/back edge 做事，那你有可能做了重複的操作而導致 WA。
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:54]: 提到看到兩遍是好事
但不太懂後面寫重複操作而WA

你可以舉例子 或是直接刪掉那句話

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+        } else if(color[v] == 1) {
+            //第一次看到的back edge
+            ...
+        }
+        //如果color[v] == 2，代表 v 的 DFS 已經結束，因此 (u,v) 這條邊已經被 v 看過了
+        //如果還沒，那 v 的 DFS 就不應該結束。
+    }
+    color[u] = 2;
+}
+```
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:91]: 不太懂 
在有像圖 就只會有 tree edge and back edge 你這邊看起來還是用 無像圖做例子 
這兩份code沒差吧？

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+<img src="image/AP.JPG" width="500" style="display:block; margin: 0 auto;"/>
+
+Bridge 指的是一張圖 \\(G \\) 移除一條邊 \\(e \\) 後 Connected Component 的數量變多，則邊 \\(e \\) 為 Bridge。中文常稱之為橋。例如下圖的紅色邊就是 Bridge
+<img src="image/Bridge.JPG" width="500" style="display:block; margin: 0 auto;"/>
+
+那我們要如何快速找到圖上所有的 AP 跟 Bridge 呢？很容易可以想到枚舉每個點或邊，把他拔掉之後看看圖上有沒有多出新的 Connected Component。但這樣做的時間複雜度會是 \\(O((V+E)^2) \\)。不過實際上，我們只要好好觀察圖上的性質就可以在 \\(O(V+E) \\) 的時間做完！以下介紹兩種不同的方法來找出圖上所有的 AP 跟 Bridge
+
+## Tarjan's Algorithm for AP/Bridge
+
+### 觀察 for AP
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:107]: 這標題怪怪的

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+<img src="image/Tarjan AP Observation.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+更一般的說，對於一個點 \\(v \\)，如果 \\(v \\) 的某一棵子樹無法在不經過 \\(v \\) 的情況下走到 \\(v \\) 的祖先，那麼 \\(v \\) 一定是 AP。
+
+<img src="image/General AP.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+但 root 是沒有祖先的，因此 root 我們要拉出來特別判斷。很明顯，當 root 有至少兩棵子樹的時候，root 一定會是 AP，否則就不是。
+<img src="image/Root AP Observation.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+### 觀察 for Bridge
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:120]: 這標題怪怪的

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+            if(low[v] > depth[u]) Bridge.emplace_back(u,v);
+        } else {
+            low[u] = min(low[u], depth[v]);
+        }
+    }
+
+    if(u == parent && child < 2) isAP = false;
+    if(isAP) AP.emplace_back(u);
+}
+```
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:187]: 在包裝得更模板一點

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+
+## Using DFS tree for Bridge/AP
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:189]: for -> to find

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+
+## BCC(Bi-connected Component)
+
+BCC 指的是沒有 AP 的 Connected Component，在中文常稱之為點雙連通分量。例如下圖中有三個 BCC
+<img src="image/Biconnected Component.JPG" width="300" style="display:block; margin: 0 auto;"/>
+
+不同的 BCC 之間會共用點，共用的部分恰好會是圖上的 AP。
+而要找出圖上所有的 BCC，我們可以通過修改找 AP 的演算法來達成。
+
+### 如何修改
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:308]: Need more explanation

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+            } while(x.first != 1)
+        }
+    }
+}
+
+```
+
+## BCC(Bridge Connected Component)
+
+BCC 指的是沒有 Bridge 的 Connected Component，在中文常稱之為邊雙連通分量、橋連通分量。例如下圖我們能找到兩個 BCC
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:363]: 怎麼又是一樣的BCC

你可以分一下 BCC-Vertex and BCC-Edge

File: src/graph/introduction_to_AP_bridge_BCC.md [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+<img src="image/Bridge Connected Component.JPG" width="300" style="display:block; margin: 0 auto;"/>
+
+不同的 BCC 沒有交集。而要找出圖上所有的 BCC，我們可以通過修改找 Bridge 的算法來達成
+
+### 如何修改
+
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge_BCC.md:369]: need more explanation

**Review by harry900831 — CHANGES_REQUESTED** [review]
- Move BCC to another md file
- please provide template code
- please add an example problem with detail explanation that can be solve by bridge/ap, BCC after you intro the respective topic.
- For the last problem section, you only need to hint the user how to solve it
- Please re-request review after you finish it.

**Comment by Lotanie** [comment]
@harry900831 助教好，我修改了助教提到的問題並補完了題目相關的部分，想請助教幫忙 review，謝謝助教!

File: src/graph/introduction_to_AP_bridge.md [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+    
+        if(u == parent && child < 2) isAP = false;
+        if(isAP) AP.emplace_back(u);
+    }
+    
+    vector<int> get_AP() {
+        return AP;
+    }
+    
+    vector<pair<int, int> > get_bridge() {
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:222]: vector<pair<int, int>> is fine.
After C++11, you don't have to add additional space between >>, please fix it.

File: src/graph/introduction_to_AP_bridge.md [hunk 12]
```diff
@@ -0,0 +1,489 @@
...
+        return Bridge;
+    }
+};
+```
+
+## Using DFS tree for Bridge/AP
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:228]: Use DFS Tree to get Bridge/AP

Comment (harry900831) [reply src/graph/introduction_to_AP_bridge.md:228]: Get Bridge/AP by DFS tree

File: src/graph/introduction_to_AP_bridge.md [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+### 觀察 Bridge 的性質
+
+我們觀察圖中那些邊絕對不可能是 bridge
+
+<img src="image/AP, Bridge/DFS Tree Observation.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+- back edge 絕對不會是 bridge。
+- **如果\\( (u,v) \\)是 back edge，那麼樹上 \\(u \\) 到 \\( v \\) 的路徑都不會是 bridge**。例如圖中因為有 \\( (F,C) \\) 這條 back edge，因此樹上 \\(F \\) 到 \\(C \\) 的路徑都不會是 bridge。
+
+所以，如果我們每遇到一條 back edge \\( (u,v) \\)，就把樹上 \\(u \\) 到 \\( v \\) 的路徑都標記成不是 bridge，那麼最後那些沒被標記到的邊就會是 bridge。問題是，**我們要如何快速標記一條路徑上所有的邊？**
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:239]: 沒被標記的點一定是 bridge 嗎？一兩句話解釋一下。

File: src/graph/introduction_to_AP_bridge.md [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+
+這個問題我們可以用前綴和的想法來解決。當我們遇到一條 back edge \\( (u,v) \\) 時，就在 \\(u \\) 上 +1，在 \\( v \\) 上 -1。 **這代表說有一條 back edge 從 \\(u \\) 開始，在 \\( v \\) 結束**。
+
+<img src="image/AP, Bridge/prefix 1.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+這樣當我們由下而上計算前綴和時，\\(u \\) 到 \\( v \\) 的 path 就全部被標記好了!
+
+<img src="image/AP, Bridge/prefix 2.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+### 前綴和 -> 子樹標記總和
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:251]: 前綴和->路徑和？

File: src/graph/introduction_to_AP_bridge.md [hunk 15]
```diff
@@ -0,0 +1,489 @@
...
+                sum[u]+=sum[v];
+            } else if(color[v] == 1) {
+                sum[u]+=1;
+                sum[v]-=1;
+            }
+        }
+        color[u] = 2;
+    }
+        
+    vector<pair<int, int> > get_bridge() {
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:302]: Ditto

File: src/graph/introduction_to_AP_bridge.md [hunk 16]
```diff
@@ -0,0 +1,489 @@
...
+        color.assign(n+1, 0);
+        sum.assign(n+1, 0);
+        backEdgeEnd.assign(n+1, 0);
+        G.assign(n+1, vector<int>());
+        AP.clear();
+    }
+    
+    void add_edge(int u, int v) {
+        G[u].emplace_back(v);
+        G[v].emplace_back(u);
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:349]: Do you test the correctness of the code?

File: src/graph/introduction_to_AP_bridge.md [hunk 17]
```diff
@@ -0,0 +1,489 @@
...
+
+```cpp
+struct bridge {
+    vector<int> color, sum;
+    vector<vector<int> > G;
+    vector<pair<int, int>> Bridge;
+    void init(int n) {
+        color.assign(n+1, 0);
+        sum.assign(n+1, 0);
+        G.assign(n+1, vector<int>());
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:272]: Do you test the correctness of the code?

File: src/graph/introduction_to_AP_bridge.md [hunk 18]
```diff
@@ -0,0 +1,489 @@
...
+
+AP 模板題
+
+> [CSES - Necessary Roads](https://cses.fi/problemset/task/2076)
+>
+> 給定一張 \\( N \\) 個點 \\( M \\) 條邊的無向圖，要你找出圖上所有的 Bridge
+
+Bridge 模板題
+
+>[Codeforce - Two Fairs](https://codeforces.com/contest/1259/problem/E)
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:404]: Problem size

File: src/graph/introduction_to_AP_bridge.md [hunk 19]
```diff
@@ -0,0 +1,489 @@
...
+
+<img src="image/AP, Bridge/Two Fair Solution 1.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+那麼答案就會是 \\(i \times j \\)，現在的問題只剩下我們要如何求出 \\(i,j\\)。
+
+要求出 \\( i \\)，我們可以拔掉 \\( a \\) 點後從 \\( b \\) 點 DFS 出去，沒被走到的點數就會是 \\( i \\)。\\( j \\) 也能用相似的方法求出
+
+</details>
+
+>[Codeforces - Bertown roads](https://codeforces.com/contest/118/problem/E)
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:427]: Problem size

File: src/graph/introduction_to_AP_bridge.md [hunk 20]
```diff
@@ -0,0 +1,489 @@
...
+
+而對於 \\(v\\) 所回到的點 \\( v' \\)，我們發現這又是一個同樣的問題，我們還是可以通過某條 back edge 向上走，重複這個過程直到回到 root。
+
+因此，任意兩點都能通過回到 root 後再由 root 走到對方，得證。
+
+</details>
+
+## Problems
+
+>[Uva - Mining Your Own Business](https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=246&page=show_problem&problem=3549)
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:453]: Problem size

File: src/graph/introduction_to_AP_bridge.md [hunk 21]
```diff
@@ -0,0 +1,489 @@
...
+
+考慮一張把原圖的 AP 全都拔掉的新圖 \\( G' \\)，會發現 \\( G' \\) 中每一個 component 最多只要塗黑 2 個點便能達成題目要求。
+
+因此可以在原圖上將這些 component 都縮成一個點，接著去討論每個 component 應該要塗黑幾個點。
+
+</details>
+
+>[Codeforces - Break Up](https://codeforces.com/problemset/problem/700/C)
+>
+> 給一張 \\( N \\) 個點 \\( M \\) 條邊的帶權無向圖與兩點 \\(S \\) , \\(T \\)，要你選至多兩條邊刪除後使 \\(S \\) , \\(T \\) 不連通。要求選的邊權重和最小。\\( N \leq 1000, M \leq 30000 \\)
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:467]: There should be an empty line between the statement and the problem size.

File: src/SUMMARY.md [hunk 22]
```diff
@@ -10,6 +10,9 @@
   - [Basic Greedy](greedy/basic.md)
 - [Dynamic Programming]()
   - [Basic DP]()
+- [Graph]()
+  - [Ap,Bridge](graph/introduction_to_AP_bridge.md)
+  - [BCC](graph/introduction_to_BCC.md)
```
Comment (harry900831) [inline src/SUMMARY.md:15]: BCC-Vertex and BCC-Edge

File: src/graph/introduction_to_AP_bridge.md [hunk 23]
```diff
@@ -0,0 +1,489 @@
+# AP, Bridge
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:1]: rename this too

File: src/graph/introduction_to_BCC.md [hunk 24]
```diff
@@ -0,0 +1,218 @@
+# BCC
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:1]: rename this too

File: src/graph/introduction_to_BCC.md [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+
+BCC-Vertex 指的是沒有 AP 的 Connected Component，在中文常稱之為點雙連通分量。例如下圖中有三個 BCC-Vertex
+<img src="image/BCC/Biconnected Component.JPG" width="300" style="display:block; margin: 0 auto;"/>
+
+BCC-Vertex 有以下幾個性質:
+
+- 不同的 BCC-Vertex 之間最多共用一個點，而該點必然是圖上的 AP
+- 一個至少有三個點的 BCC-Vertex，任兩點間至少會有兩條沒有共用邊的簡單路徑
+- 對於在同一個 BCC-Vertex 中的任意三個相異點 \\(a,b,c\\)，必存在一條簡單路徑依序經過 \\(a,b,c\\) 三點
+- 把每個 BCC-Vertex 縮成一個點，他周圍的點會是這個 BCC-Vertex 中的 AP。
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:17]: 每一點都需要一兩句話稍微解釋一下

Comment (harry900831) [reply src/graph/introduction_to_BCC.md:17]: > 把每個 BCC-Vertex 縮成一個點，他周圍的點會是這個 BCC-Vertex 中的 AP。

這句話看不懂

File: src/graph/introduction_to_BCC.md [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+
+而要如何找出圖上所有的 BCC-Vertex 呢?我們發現:
+
+- 對於那些不是 AP 的點，可以通過 DFS 找出所有與他同屬一個 BCC-Vertex 的點(當我們遇到 AP 時就不要再 DFS 下去即可)
+- 對於 AP，只要檢查他周圍的點有沒有 AP，有的話這兩個 AP 就會形成一個 BCC-Vertex。
+
+例如下圖中，\\( A \\) 通過 DFS 可以找到 \\( B,C,D \\)，這四點恰為一個 BCC-Vertex。\\( A \\) 點不會找到 \\( E \\)，因為 \\( D \\) 點是 AP，不會繼續 DFS 下去。而 \\ (D \\) 點會檢查到 \\ (E \\) 是 AP，因此 \\(D,E\\) 會是一個 BCC-Vertex。
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:24]: Render error

File: src/graph/introduction_to_BCC.md [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+
+題目要找 AP, Bridge, BCC-Vertex 的數量，以及擁有最多邊的 BCC-Vertex 的邊數。
+
+其實就是模板題
+
+</details>
+
+>[Codeforce - Simple Cycles Edges](https://codeforces.com/problemset/problem/962/F)
+>
+>給定一張 \\( N \\) 個點 \\( M \\) 條邊的無向圖，問那些邊剛好位於一個簡單環上。
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:99]: Problem size

File: src/graph/introduction_to_BCC.md [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+        if(depth[v] < depth[u]) stk.emplace(u,v);
+        if(depth[v] == 0) {
+            dfs(v, u, dep+1);
+            low[u] = min(low[v], low[u]);
+            if(low[v] >= depth[u]) {
+                edge x;
+                bcc.emplace_back({});
+                do {
+                    x = stk.top(); stk.pop();
+                    bcc.back().emplace_back(x);
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:61]: Wrap it as a struct such as:
https://github.com/OmeletWithoutEgg/ckiseki/blob/master/codes/Graph/BCC_Vertex.cpp

File: src/graph/introduction_to_BCC.md [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+
+## BCC-Edge(Bridge Connected Component)
+
+BCC-Edge 指的是沒有 Bridge 的 Connected Component，在中文常稱之為邊雙連通分量、橋連通分量。例如下圖我們能找到兩個 BCC-Edge
+<img src="image/BCC/Bridge Connected Component.JPG" width="300" style="display:block; margin: 0 auto;"/>
+
+BCC-Edge 有以下幾個性質
+
+- 同一個 BCC-Edge 的任兩點間至少存在兩條沒有共用邊的簡單路徑
+- 如果把每個 BCC-Edge 縮成一個點，那麼新得到的圖會是一棵樹或者森林。
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:117]: Ditto, 需要幾句話解釋一下

File: src/graph/introduction_to_BCC.md [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+        } else {
+            low[u] = min(low[u], depth[v]);
+        }
+    }
+    
+    if(low[u] == depth[u]) {
+        bcc.emplace_back({});
+        int x;
+        do {
+            x = stk.top(); stk.pop();
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:158]: Ditto

File: src/graph/introduction_to_BCC.md [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+    }
+}
+
+```
+
+## Exercise
+
+>[2015 ACM Amman Collegiate Programming Contest pH - Bridges](https://codeforces.com/gym/100712)
+>
+> 請有興趣的讀者自行去連結查看題目敘述，以免暴雷未來打算模擬這場比賽當作團練的人。
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:170]: 這不是台北站 應該可以放題目敘述

File: src/graph/introduction_to_BCC.md [hunk 32]
```diff
@@ -0,0 +1,218 @@
...
+
+我們首先把所有的 BCC-Edge 縮點，根據 BCC-Edge 的性質我們會得到一棵樹，樹上的邊都是 bridge。
+
+我們思考如果在樹上 \\( u,v \\) 兩點間加上一條邊後，有哪些邊就不再是 bridge 了。可以發現樹上從 \\( u \\) 到 \\( v \\) path 上的所有邊都不再是 bridge。而我們想要讓最多的 bridge 消失，因此問題就轉換為了求樹的直徑。
+
+</details>
+
+## Problems
+
+>[TIOJ - 圓桌武士](https://tioj.ck.tp.edu.tw/problems/1684)
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:182]: Problem size

File: src/graph/introduction_to_BCC.md [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+>
+>有 \\( N \\) 個武士和 \\( M \\) 個武士之間倆倆的仇恨關係，如果能選出奇數個武士(至少三個)坐在圓桌邊，且相鄰的兩人不互相仇恨，那他們就可以開會。問有多問有多少位武士永遠不可能開會。
+
+<details><summary> Hint </summary>
+
+反過來思考，我們改成找可以和別人開會的武士有誰。我們可以觀察原圖的補圖 (即武士之間的不仇恨關係)，想想看這張補圖上符合怎樣性質的點至少可以參與到一場會議中。
+
+</details>
+
+>[Codeforces - Tourist Reform](https://codeforces.com/contest/732/problem/F)
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:192]: Problem size

File: src/graph/introduction_to_BCC.md [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+>
+>給定一張 \\( N \\) 個點 \\( M \\) 條邊的無向圖，定義 \\(r_i \\) 為幫每條邊定向後，\\(i \\) 點可以走到的點數。要你給出一種定向方法使得 \\(min_i({r_i})\\) 最大
+
+<details><summary> Hint </summary>
+
+找出所有的 BCC-Edge 後，想想看最大化的 \\(min_i({r_i})\\) 至少會等於什麼?
+
+</details>
+
+> [POJ - Road Construction](http://poj.org/problem?id=3352)
```
Comment (harry900831) [inline src/graph/introduction_to_BCC.md:202]: Problem size

**Review by harry900831 — CHANGES_REQUESTED** [review]
- Overall, LGTM
- No space in filename and directory name, please fix it.
- Please fix textlint
- I might merge it after your next revision, please check the whole article by yourself in advance.

File: src/graph/introduction_to_AP_bridge.md (hunk 16 above)
Comment (Lotanie) [reply src/graph/introduction_to_AP_bridge.md:349]: Yes

File: src/graph/introduction_to_AP_bridge.md (hunk 17 above)
Comment (Lotanie) [reply src/graph/introduction_to_AP_bridge.md:272]: Yes

File: src/graph/introduction_to_AP_bridge.md [hunk 35]
```diff
@@ -0,0 +1,489 @@
...
+
+我們觀察下圖中 \\( C \\) 點的左子樹。你會發現這棵子樹一定要通過 \\( C \\) 點才能走到 \\( C \\) 點的祖先，所以理所當然的，當我們將 \\( C \\) 點移除，整張圖就會斷成兩個 Connected Component。而 \\( C \\) 點就會是 AP。
+
+<img src="image/AP, Bridge/Tarjan AP Observation.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+更一般的說，對於一個點 \\(v \\)，如果 \\(v \\) 的某一棵子樹無法在不經過 \\(v \\) 的情況下走到 \\(v \\) 的祖先，那麼 \\(v \\) 一定是 AP。
+
+<img src="image/AP, Bridge/General AP.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+但 root 是沒有祖先的，因此 root 我們要拉出來特別判斷。很明顯，當 root 有至少兩棵子樹的時候，root 一定會是 AP，否則就不是。
```
[the remark this answers is not in the document]
Comment (Lotanie) [reply src/graph/introduction_to_AP_bridge.md:126]: 想問這部分是不是可以放在 DFS tree 的章節一起講，然後這邊就只要簡單提到說我們在 DFS tree 有說明過為何不會有橫跨子樹的 edge 這樣？

[the remark this answers is not in the document]
Comment (harry900831) [reply src/graph/introduction_to_AP_bridge.md:126]: 其實我有在想，你這邊介紹滿多DFS Tree，感覺幾乎快把DFS Tree的內容講完了。理想上感覺應該是把你前半部分DFS Tree 的內容移到 DFS_Tree.md ，並補全DFS Tree的章節，不過我不會要求你這樣做(如果你願意我會很開心，不一定要在這次PR做)。

[the remark this answers is not in the document]
Comment (harry900831) [reply src/graph/introduction_to_AP_bridge.md:126]: 那你這邊就一兩句話提一下說不會有橫跨子樹的 edge，讀者可以簡單想一下，那條 edge一定會變成tree edge 之類的

**Comment by Lotanie** [comment]
@harry900831 助教好，我修改了助教所提到的問題，想請助教幫忙 review，謝謝助教!

File: src/graph/introduction_to_AP_bridge.md [hunk 36]
```diff
@@ -0,0 +1,534 @@
...
+DFS tree 就如同他的名字，是一棵根據 DFS 的過程長出的 tree。由於其良好的性質，在處理圖相關的問題時是很有用的工具。
+
+考慮以下 DFS 的走訪，那些直接走訪的邊被稱為 tree edge，而其他的邊被稱為 back edge。
+
+```cpp
+void DFS(int u) {
+    visited[u] = true;
+    for(int &v : G[u]) {
+        if(!visited[v]) {
+            DFS(v);//u->v 為 tree edge
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:38]: move the comment above this line

File: src/graph/introduction_to_AP_bridge.md [hunk 37]
```diff
@@ -0,0 +1,534 @@
...
+            //第一次看到的back edge
+            //在這個例子中當 E 看到 (E,A) 這條邊時會進到這裡
+            backedgeCounter[u]+=1;
+            backedgeCounter[v]+=1;
+        }
+        //在這個例子中，當 A 看到 (E,A) 這條邊時
+        //因為 E 的 DFS 已經結束， color[E] = 2
+        //因此 (E,A) 這條邊不會被算到第 2 次
+    }
+    color[u] = 2;
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:100]: // 和註解文字 要空一個空白

Comment (harry900831) [reply src/graph/introduction_to_AP_bridge.md:100]: 請針對這對所有做修正

File: src/graph/introduction_to_AP_bridge.md [hunk 38]
```diff
@@ -0,0 +1,534 @@
...
+
+<img src="image/AP_and_Bridge/Tarjan_AP_observation.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+更一般的說，對於一個點 \\(v \\)，如果 \\(v \\) 的某一棵子樹無法在不經過 \\(v \\) 的情況下走到 \\(v \\) 的祖先，那麼 \\(v \\) 一定是 AP。
+
+<img src="image/AP_and_Bridge/General_AP.JPG" width="400" style="display:block; margin: 0 auto;"/>
+
+但 root 是沒有祖先的，因此 root 我們要拉出來特別判斷。很明顯，當 root 有至少兩棵子樹的時候，root 一定會是 AP，否則就不是。
+
+<img src="image/AP_and_Bridge/Root_AP_Observation.JPG" width="400" style="display:block; margin: 0 auto;"/>
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:130]: 圖片上不要有字體很醜的中文 可以改英文

File: src/graph/introduction_to_AP_bridge.md [hunk 39]
```diff
@@ -0,0 +1,534 @@
...
+};
+```
+
+### 如何找 AP
+
+我們首先回想一下 Tarjan 對 AP 的觀察：
+
+- 對於 root，如果他有至少兩棵子樹的時候，他就一定是 AP，否則就不是。
+- 對於一個點 \\(v \\)，如果 \\(v \\) 的某一棵子樹無法在不經過 \\(v \\) 的情況下走到 \\(v \\) 的祖先，那麼 \\(v \\) 一定是 AP。
+<img src="image/AP_and_Bridge/General AP.JPG" width="400" style="display:block; margin: 0 auto;"/>
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:343]: render error

File: src/graph/introduction_to_AP_bridge.md [hunk 40]
```diff
@@ -0,0 +1,534 @@
...
+
+第一點很好解決。
+
+對於第二點，我們要檢查的其實就是**對於一棵子樹，有沒有 back edge 跨過 \\( v \\)**，如果沒有那 \\( v \\) 就會是 AP。
+
+而這個東西要計算的其實就是：
+
+- \\((v\ 的子樹標記總和)\\) \\(-\\) \\((\\)這顆子樹在 \\( v \\) 結束的 back edge 數量\\()\\)。
```
Comment (harry900831) [inline src/graph/introduction_to_AP_bridge.md:351]: why 前半部份要包在mathjax內？

**Review by harry900831 — CHANGES_REQUESTED** [review]
LGTM, I'll merge it after your revision.

**Comment by Lotanie** [comment]
@harry900831 助教好，我修改了助教所提到的問題，想請助教幫忙 review，謝謝助教!

**Comment by harry900831** [comment]
Thanks for your contribution!

**Pull request merged**